### ISSUE TITLE
FreeTypeFontGenerator produces very different result compared to Hiero

### ISSUE THREAD

**Issue by MontyHimself** [libgdx/libgdx#6547]
#### Issue details
I am trying to use the `FreeTypeFontGenerator` to generate a `BitmapFont` for the Calibri typeface. This is the code I use for generation:

```java
gridCoordFonts = new BitmapFont[4];
FreeTypeFontGenerator generator = new FreeTypeFontGenerator(Gdx.files.internal("assets/fonts/calibri/calibri.ttf"));
for (int i = 0; i < baseCamera.getZoomLevels().length; i++) {
    FreeTypeFontGenerator.FreeTypeFontParameter parameter = new FreeTypeFontGenerator.FreeTypeFontParameter();
    parameter.size = Math.round(16f / baseCamera.getZoomLevels()[i]);
    parameter.renderCount = 1;
    parameter.padTop = 1;
    parameter.padBottom = 1;
    parameter.padLeft = 1;
    parameter.padRight = 1;
    BitmapFont font = generator.generateFont(parameter);
    gridCoordFonts[i] = font;
}
generator.dispose();
```

The result looks much different than when I generate the `BitmapFont` through Hiero (with the default parameters) and load it in as follows:

```java
Texture fontTexture = new Texture(Gdx.files.internal("assets/calibri16.png"), true);
fontTexture.setFilter(Texture.TextureFilter.Nearest, Texture.TextureFilter.Nearest);
font = new BitmapFont(Gdx.files.internal("assets/calibri16.fnt"), new TextureRegion(fontTexture), false);
```

I tried to play with the hinting parameter, but it doesn't change the output to that of the Hiero case.

With `FreeTypeFontGenerator`:

![freetype_generator](https://user-images.githubusercontent.com/16304918/120214743-2e410380-c235-11eb-9dde-7773d2ea1466.png)

With Hiero:

![freetype_hiero](https://user-images.githubusercontent.com/16304918/120214758-313bf400-c235-11eb-883c-84e8845e728d.png)

#### Version of LibGDX and/or relevant dependencies
libgdx 1.10.0
gdx-freetype 1.10.0

#### Please select the affected platforms
- [ ] Android
- [ ] iOS
- [ ] HTML/GWT
- [X] Windows
- [ ] Linux
- [ ] MacOS

**Comment by MontyHimself** [comment]
I just realized that there is an option in Hiero to choose the renderer to generate the bitmaps. When I set that to FreeType (instead of Java, which is the default), Hiero doesn't render the font at all (the preview window is just black). I tried it with the Ubuntu font, and that works fine in both modes.

Maybe the issue is that the freetype version which both the gdx-freetype extension and Hiero use cannot render the Calibri TTF for some reason, and then the `FreeTypeFontGenerator` falls back to some other default font face?

**Comment by tommyettinger** [comment]
I'm investigating this, and there's some definitely-strange behavior regarding hinting. I can use FreeType as an option in my version of Hiero, but it certainly looks different from Java or Native rendering. The ASCII and Latin-1 characters (the first 256) are shown with really awful, harsh hinting, but the later characters look fine.

![See this preview](https://i.imgur.com/DN3TFcD.png)

I suspect this is related to the `incremental` flag internal to FreeType, or maybe something else that could render later glyphs differently from earlier ones. I tried changing the hinting settings in the source code, but it doesn't seem to change the hinting for the first 256 glyphs.

This does seem like a bug, and it's unlikely that Calibri (as one of the MS-blessed fonts) is a faulty file. I don't know much about FreeType in libGDX, as I have never used the extension, so this might be something for @NathanSweet when he has the time.

In the meanwhile, you can use [BMFont](https://angelcode.com/products/bmfont/), which has options to disable hinting or configure smoothing, if you want to make a font that uses the same style of hinting for all codepoints.

**Comment by NathanSweet** [comment]
I'm not seeing initial glyphs (caps below) being different from later glyphs (lowercase below) when using FreeType to render Calibri in Hiero:
![](http://n4te.com/x/1901-Xlmj.png)
If you are seeing a difference for some reason, you can use the reset cache button to recreate the font with all the glyphs up front.

The Java renderer is a bit blurrier.

Sorry to say I don't really know what is going on. I no longer use Hiero personally, instead I write code that generates fonts with FreeType. That way I can easily customize it and it's easier to get the same results later. I also have fonts generated as part of a build process.

**Comment by tommyettinger** [comment]
In the screenshot I posted with the gray background, the ASCII and Latin-1 sections (up to char `\u00FF`) get very strong hinting, and later codepoints get normal rendering, like the Java renderer. Just ASCII never showed a problem for me. I did try resetting the cache, with the same result.

I should probably mention that I pretty much only use Hiero to attempt to fix bugs other people are encountering, rather than to actually make fonts myself. I usually use either [BMFont](https://angelcode.com/products/bmfont/) or [a hackish tool I wrote that can generate SDF and MSDF fonts](https://github.com/tommyettinger/Glamer). I'm not really sure if it's worth it to continue maintaining Hiero, personally.

**Comment by cordelleus** [comment]
How would you test the FreeTypeFontGenerator with the LibGDX3 graphics buffer?  That might fix the issue.  The most important thing to consider is that you absolutely need to limit and then reduce the number of ASCII characters available to your game.  This might sound superfluous and over the top for my next suggestion, but you can limit the enflow of what characters your game allows for your ASCII character selection.  For instance if you wanted to go crazy with it, you can limit all character to just A through Z and forbid 0 through 9 for example.  This is some thing that LibGDX would allow.  You would not want to do that in Java for the operating system it self.  That would cause problems with the cross platform nature of Java.  LibGDX handles it better than Java alone.  I recommend looking at Android references for limiting ASCII character if you need help or assistance with your coding attempts.  Essentially, you can also product a similar type of result by editing a type of ".FNT" file I think.

Think about it:  Hiero references graphics to ASCII codes.  Just edit the ASCII codes available them selves.  LibGDX should be intelligent enough to skip over font character references if necessary.  Android seems capable enough in this direction at least.  I would say to start there, if it helps you.  Remember!  You are editing the ASCII additions that your game allows.  This has no thing to do with the graphics.

@NathanSweet I apologize again for my previous rudeness behavior.  I am trying to cleanse COVID nonsense, where I am.  I erased one source due to MILSPEC weaponized things.  Would you help with the .FNT file thing, please?  I am not able to assist with my compiler tools at the moment.  IntelliJ Community is not available to me right now.  Thank you.

**Comment by terefang** [comment]
maybe this is a bug or feature in freetype itself ?

**Comment by Frosty-J** [comment]
Yeah, I think it's just a FreeType thing. I've been here before after installing Microsoft fonts.

https://ask.libreoffice.org/t/calibri-font-not-rendering-correctly/2079/2
